pkg/apis/flowcontrol/validation/validation.go

```diff
@@ -0,0 +1,265 @@
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("nonResourceURLs"), rule.NonResourceURLs, "if '*' is present, must not specify other non-resource URLs"))
+		}
+		return allErrs
+	}
+	if len(rule.APIGroups) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("apiGroups"), "resource rules must supply at least one api group"))
+	}
+	if len(rule.Resources) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("resources"), "resource rules must supply at least one resource"))
+	}
+
+	if hasWildcard(rule.Verbs) && len(rule.Verbs) > 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("verbs"), rule.Verbs, "if '*' is present, must not specify other verbs"))
+	}
+	if hasWildcard(rule.APIGroups) && len(rule.APIGroups) > 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("apiGroups"), rule.APIGroups, "if '*' is present, must not specify other api groups"))
+	}
+	if hasWildcard(rule.Resources) && len(rule.Resources) > 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("resources"), rule.Resources, "if '*' is present, must not specify other resources"))
+	}
+	if !supportedVerbs.IsSuperset(sets.NewString(rule.Verbs...)) {
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("verbs"), rule.Verbs, supportedVerbs.List()))
+	}
+
+	return allErrs
+}
+
+// ValidateFlowSchemaStatus validates status for the flow-schema.
+func ValidateFlowSchemaStatus(status *flowcontrol.FlowSchemaStatus, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	// conditions will not be validated
+	return allErrs
+}
+
+// ValidatePriorityLevelConfiguration validates priority-level-configuration.
+func ValidatePriorityLevelConfiguration(pl *flowcontrol.PriorityLevelConfiguration) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMeta(&pl.ObjectMeta, false, ValidatePriorityLevelConfigurationName, field.NewPath("metadata"))
+	allErrs = append(allErrs, ValidatePriorityLevelConfigurationSpec(&pl.Spec, pl.Name, field.NewPath("spec"))...)
+	allErrs = append(allErrs, ValidatePriorityLevelConfigurationStatus(&pl.Status, field.NewPath("status"))...)
+	return allErrs
+}
+
+// ValidatePriorityLevelConfigurationSpec validates priority-level-configuration's spec.
+func ValidatePriorityLevelConfigurationSpec(spec *flowcontrol.PriorityLevelConfigurationSpec, name string, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if !spec.Exempt {
+		if spec.AssuredConcurrencyShares <= 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("assuredConcurrencyShares"), spec.AssuredConcurrencyShares, "must be positive"))
+		}
+		if spec.QueueLengthLimit <= 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("queueLengthLimit"), spec.QueueLengthLimit, "must be positive"))
+		}
+		if spec.Queues <= 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("queues"), spec.Queues, "must be positive"))
+		}
+		if spec.HandSize < 0 {
```

Comment:
what does a zero hand size mean?

Reply:
Defaulting logic replaces zero for a non-exempt priority level, and hand size is meaningless for an exempt level.